diff/order.go

```diff
@@ -0,0 +1,84 @@
+package diff
+
+import "github.com/kong/deck/types"
+
+/*
+                                       Root
+                                         |
+         +----------+----------+---------+------------+---------------+
+         |          |          |         |            |               |
+         v          v          v         v            v               v
+L1    Service    RbacRole  Upstream  Certificate  CACertificate  Consumer ---+
+      Package        |         |        |     |      |                |      |
+        |            v         v        v     |      v                v      |
+L2      |        RBACRole   Target     SNI    +-> Service       Credentials  |
+        |        Endpoint                         |  |              (7)      |
+        |                                         |  |                       |
+        |                                         |  |                       |
+L3      +---------------------------> Service <---+  +-> Route               |
+        |                             Version        |     |                 |
+        |                                 |          |     |                 |
+        |                                 |          |     v                 |
+L4      +----------> Document   <---------+          +-> Plugins  <----------+
+*/
+
+var dependencyOrder = [][]string{
+	{
+		types.ServicePackage,
+		types.RBACRole,
+		types.Upstream,
+		types.Certificate,
+		types.CACertificate,
+		types.Consumer,
+	},
+	{
+		types.RBACEndpointPermission,
+		types.Target,
+		types.SNI,
+		types.Service,
+
+		types.KeyAuth, types.HMACAuth, types.JWTAuth,
+		types.BasicAuth, types.OAuth2Cred, types.ACLGroup,
+		types.MTLSAuth,
+	},
+	{
+		types.ServiceVersion,
+		types.Route,
+	},
+	{
+		types.Plugin,
+		types.Document,
+	},
```

Comment:
Note to reviewers, please then verify if the above ascii is reflected in this data structure correctly.
I took all entities at the same level and put them together.

Reply:
I think this DS reflect the vertical relationship, the dependencies relationship is not that clear, how do you think ?

Reply:
Yes and no--this correctly reflects the levels, but the levels alone aren't the dependency relationship, and this isn't actually a graph.

Whether that matters or not is "probably not" though, since simply grouping things into levels is sufficient for achieving the goal of the refactor, i.e. that we no longer use an absolute order for all entities and waits, and instead simply place entities into a group below all their required parents, with waits between the groups.

I don't see anything that an actual DAG would get us over this approach for that need, though it may have other uses--in the KIC context a per-entity DAG would provide us more rich status information and allow us to selectively drop entities and their children from the config source (e.g. do not create a route from an Ingress when its target Service is not present, mark the Ingress unsatisfied because the Service is missing). However, that's a much more radical change because it's entity relationships rather than entity class relationships, and is more complicated to implement (we'd need an actual graph type and BFS that waits between the levels).

I don't think there'd be much benefit to a graph with the entity class approach yet. It'd mean we could handle, say, credential creation independent of upstream creation, but in practice our graph is simple enough at present (and unlikely to get hugely more complex) that I think the simple level approach is fine. It may matter because of type cardinality: for example, if you have a very large number of consumers checking all those blocks creation of new targets, though I don't know whether that has a meaningful UX impact--that block _already existed_ in the approach prior to this refactor, and I don't think anyone was complaining about it.

Reply:
> I think this DS reflect the vertical relationship, the dependencies relationship is not that clear, how do you think ?

This DS doesn't track the dependency relationships because that is not required for execution. We simplify the graph into levels as that is enough information to correctly execute.